### ISSUE TITLE
Error: PHP Notice: Undefined offset: 74 in /usr/local/wpcs/WordPress/Sniffs/WhiteSpace/ControlStructureSpacingSniff.php on line 107

### ISSUE THREAD

**Issue by PC888** [squizlabs/PHP_CodeSniffer#1087]
Hey guys, I've just started receiving this error when editing php files. 

Error: PHP Notice: Undefined offset: 74 in /usr/local/wpcs/WordPress/Sniffs/WhiteSpace/ControlStructureSpacingSniff.php on line 107
    at ChildProcess.<anonymous> (/Users/paulcarter/.atom/packages/linter-phpcs/node_modules/atom-linter/node_modules/sb-exec/lib/index.js:36:20)
    at emitTwo (events.js:87:13)
    at ChildProcess.emit (events.js:172:7)
    at maybeClose (internal/child_process.js:818:16)
    at Socket.<anonymous> (internal/child_process.js:319:11)
    at emitOne (events.js:77:13)
    at Socket.emit (events.js:169:7)
    at Pipe._onclose (net.js:469:12)

I'm running linter-phpcs/linter with phpcs and the wpcs coding standards. It's been running fine since the fresh install 1 week ago.

When I disable linter-phpcs the errors stop.

I reported the error at - https://github.com/AtomLinter/linter-phpcs/issues/175. They responded and closed the ticket.

> Sounds like a PHPCS error to me. Some sniffs (especially custom/3rd-party sniffs) tend to error out when you have a syntax error in your code, or sometimes for other unknown reasons.
I'd recommend updating CodeSniffer and your ruleset and see if it solves anything, and otherwise report the error to wpcs together with the code to reproduce the error :)

I alos reported the error at - https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards

> I get the errors sometimes, they are normally caused by php code which has a syntax error in it. I normally just close the error and have no negative effects. The issue may be better reported at AtomLinter/linter-phpcs as the issue only happens in Atom and not the command line.

I'm using the 'WordPress' coding standards 'straight out of the box'; no custom sniffs or configuration.

By forcing a syntax error, the error message appears, top RHS, then as you continue to enter code, multiple error messages appear, until you're forced to click 'close all' in order to see the window again.

On force closing the errors, there doesn't seem to be any other associated negative effects.

**Comment by gsherwood** [comment]
I'm not sure what I can do with this error report. The error is coming from a WordPress coding standard sniff and I don't have any code to replicate a problem to see if it is a core PHPCS issue.

I know you've reported it to the sniff developer, so I'm wondering why you've also reported it here. Are you able to replicate an error with the core sniffs as well?

**Comment by PC888** [comment]
Thanks Greg, I'm new to Atom and PHP, so I'm still trying to get my head around how everything works, and what programs are responsible for what. I reported here as well because I don't really know who I should be talking to. I have the basic install, I followed - https://rezzz.com/php-codesniffer-with-wordpress-coding-standards-and-atom/ with no modifications, and it has been working fine for a week, so I'm not sure why I'm getting errors suddenly.

I just opened a brand new file demo.php and typed - if

Immediately, I get this error - 

Error: PHP Notice: Undefined offset: 58 in /usr/local/wpcs/WordPress/Sniffs/WhiteSpace/ControlStructureSpacingSniff.php on line 107
Error: PHP Notice:  Undefined offset: 58 in /usr/local/wpcs/WordPress/Sniffs/WhiteSpace/ControlStructureSpacingSniff.php on line 107
    at ChildProcess.<anonymous> (/Users/paulcarter/.atom/packages/linter-phpcs/node_modules/atom-linter/node_modules/sb-exec/lib/index.js:36:20)
    at emitTwo (events.js:87:13)
    at ChildProcess.emit (events.js:172:7)
    at maybeClose (internal/child_process.js:818:16)
    at Socket.<anonymous> (internal/child_process.js:319:11)
    at emitOne (events.js:77:13)
    at Socket.emit (events.js:169:7)
    at Pipe._onclose (net.js:469:12)

This doesn't seem to be a 'WP standards' issue, if it's faulting on an if statement. However, I just reset linter-phpcs code standard to the default PSR2 and I can't replicate the error, so it looks like it's with wpcs.

Appreciate your help.

![error](https://cloud.githubusercontent.com/assets/20432991/17120725/b8fd7050-5311-11e6-9591-4f4c9c3a1104.png)

**Comment by gsherwood** [comment]
I just ran every sniff that is included with PHPCS over a file containing just '<?php if' and no errors were produced, so it seems like this is just a case of the WordPress sniff needing some extra checking.

Linters that check constantly while you type are 100% guaranteed to be checking syntax errors, so I've had to make the core of PHPCS, and many sniffs, deal with these cases over the last few versions, but I'm not aware of any current issues.

I'll close this now and let you continue following up with the WordPress standard. Hopefully you've got enough information to help you find a solution, but I can confirm that just closing those error message boxes is not going to cause any problems for you in the meantime.

**Comment by PC888** [comment]
Cheers Greg, much appreciated.

Paul

**Comment by PC888** [comment]
Hey Greg, I spent a lot of time trying to sort this out with the wpcs developer yesterday. Eventually, adding a line of code to ControlStuctureSpacingSniff.php solved the (if) error - 

> Error:PHP Notice: Undefined offset: 58 in /usr/local/wpcs/WordPress/Sniffs/WhiteSpace/ControlStructureSpacingSniff.php on line 107

This was after completely uninstalling/reinstalling Homebrew, PHP CodeSniffer, wpcs and linter-phpcs.

Then while tooling around with some deliberate syntax errors (mixing single/double quotation), I got another error - 

> Error:PHP Notice:  Undefined offset: 71 in /usr/local/Cellar/php-code-sniffer/2.6.2/CodeSniffer/Standards/Squiz/Sniffs/Strings/DoubleQuoteUsageSniff.php on line 72

This one in the Code Sniffer standards! At this point the wpcs developer pulled the pin, as it wasn't a wpcs issue.

> @PC888 Ok, thanks for testing.
> Regarding the new "issue":
This is not a WPCS issue, but would need to be solved upstream in PHPCS.
I believe it's caused by the mismatched quotes in your code, so if you change echo 'It Worked"; to echo 'It Worked';, the issue should be gone.

![error](https://cloud.githubusercontent.com/assets/20432991/17161083/a83546f8-53ed-11e6-9fe4-ee763b20cf52.jpg)

Closing the error boxes is not productive. In some cases, after the error is triggered,  every subsequent keystroke, draws a new error box, until you are forced to stop and close them all.

Short of permanently disabling linter-phpcs, I don't know where else to go with this

**Comment by gsherwood** [comment]
> Short of permanently disabling linter-phpcs, I don't know where else to go with this

This is a completely different error caused by a completely different sniff. This sniff is part of the core, so it is a core bug that I will fix.

If you keep making syntax errors in your code and you have the linter configured to run constantly, you're going to potentially run into issues (as I outlined in my previous comment). If you are happy with that, you can keep reporting them to whatever project contains the sniff with the problem.

But you do need to remember that the sheer number of possible syntax errors a developer can make is enormous. Many of the common ones you encounter as you type are probably caught and fixed already, but others may not be. If this bothers you, disable the auto-linting and invoke the linter only on save, or only when you ask it to. Then, at least, you know that any syntax errors are ones that you have missed.

**Comment by gsherwood** [comment]
^ That commit fixes this particular error.

**Comment by PC888** [comment]
Thanks Greg, I downloaded the updated PHP_CodeSniffer-master.zip and replaced my local  CodeSniffer folder entirely. So far so good.

Now whilst editing my basic demo.php file I got this -

![error](https://cloud.githubusercontent.com/assets/20432991/17163498/5df67e0c-5403-11e6-9e09-ae11a9b1fc31.jpg)

**Comment by gsherwood** [comment]
> Now whilst editing my basic demo.php file I got this

Have you tried running PHPCS outside of atom? Can you post the contents of that file as text so I can try it?

**Comment by PC888** [comment]
I haven't Greg; I have no experience with PHPCS outside of Atom; and very little within for that matter.

The file I'm editing is demo.php in a basic non-wordpress sandbox.dev local site I built to test PHP while I learn. The site uses Composer to handle 'Kint' and 'Whoops' (Error/Debug Dependencies).

The contents of the file when the error appeared is below. I'm working on arrays ATM.
[demo.txt](https://github.com/squizlabs/PHP_CodeSniffer/files/385279/demo.txt)

**Comment by gsherwood** [comment]
> I haven't Greg; I have no experience with PHPCS outside of Atom; and very little within for that matter.

That's fine, I can check it with the content you've given me.

**Comment by gsherwood** [comment]
When I check that file using every available sniff, I get `Time: 107ms; Memory: 8Mb`, so I don't know where your memory blowout would be coming from.

I also tested with all the WordPress coding standards and they maxed out at 6Mb of memory.

I really have no idea what the Atom plugin could be doing to get PHPCS to max out on memory. I've not seen that before unless there were obvious PHP errors before the memory error.

The only thing you can really do is run PHPCS directly on the file and see if there is a problem. You must have already installed PHPCS so Atom can use it. Do you know where the command is? Maybe the command is already in your path and you can just run `phpcs`.

What you'll want to try doing on the command line is something like this: `phpcs /path/to/demo.php --standard=PEAR`

That will check your code against the PEAR coding standard, which is a medium-sized standard. It will confirm that the file can be parsed correctly. If you don't see any PHP errors, you'll get full PHPCS error report. You should see this:
```
$ phpcs /path/to/demo.php --standard=PEAR

FILE: /path/to/demo.php
---------------------------------------------------------------------------------------------------------
FOUND 13 ERRORS AND 1 WARNING AFFECTING 8 LINES
---------------------------------------------------------------------------------------------------------
  5 | ERROR   | [x] Tag value indented incorrectly; expected 1 spaces but found 3
  6 | ERROR   | [ ] The tag in position 2 should be the @author tag
  6 | ERROR   | [x] Tag value indented incorrectly; expected 3 spaces but found 5
  7 | ERROR   | [ ] Content of the @author tag must be in the form "Display Name <username@example.com>"
  7 | ERROR   | [x] Tag value indented incorrectly; expected 2 spaces but found 4
  8 | ERROR   | [ ] The tag in position 4 should be the @license tag
  8 | ERROR   | [x] Tag value indented incorrectly; expected 4 spaces but found 6
  9 | ERROR   | [x] Tag value indented incorrectly; expected 1 spaces but found 3
 10 | WARNING | [ ] PHP version not specified
 10 | ERROR   | [ ] Missing @category tag in file comment
 33 | ERROR   | [x] Space after opening parenthesis of function call prohibited
 33 | ERROR   | [x] Expected 0 spaces before closing bracket; 1 found
 35 | ERROR   | [x] Space after opening parenthesis of function call prohibited
 35 | ERROR   | [x] Expected 0 spaces before closing bracket; 1 found
---------------------------------------------------------------------------------------------------------
PHPCBF CAN FIX THE 9 MARKED SNIFF VIOLATIONS AUTOMATICALLY
---------------------------------------------------------------------------------------------------------

Time: 90ms; Memory: 4.75Mb
```

**Comment by PC888** [comment]
Cheers Greg, I'll have another look tomorrow. I used Homebrew to install PHPCS and I'm sure I read somewhere that with PHPCS installed globally, I could run it over a file from the command line. I'll try it out in the morning with 'a clear head'. Apologies for the time spent on this.

**Comment by gsherwood** [comment]
> Apologies for the time spent on this.

No problem at all. Let me know how you go.

**Comment by PC888** [comment]
Hey Greg, I ran all of the standards over that file and received similar results to yours, so I have no idea what else could be causing that error. It appeared once again last night; I just cleared it and kept going. [phpcs_tests.txt](https://github.com/squizlabs/PHP_CodeSniffer/files/387337/phpcs_tests.txt)

I caught a couple of new errors last night also, mostly to do with incorrect bracket matching; I just cleared those also, and kept moving. As you say, the errors are not causing any harm, so just close them.

I guess my problem is that I don't fully understand the environment. 

- It was my understanding that using Atom/PHPCS/linter-phpcs with wpcs would help me to write code that adhered to whatever coding standard I set in linter-phpcs. Hence the Error/Warning messages at the bottom of the editor; as featured in that 'package/plugin'. That's what I 'signed-up' for. It's great, they're just guidelines, but don't actually stop you from writing whatever you want.

- Where I'm coming unstuck is when these large error panels appear. To me it appears like there is a critical fault with the program/package/editor/configuration. These errors force you to stop writing and clear/acknowledge them, but the contents of the messages makes no sense to the user, unless you're the developer that wrote the programs. Apart from the memory blowout(freak occurence?), the errors have been all syntax related. We fixed the 'quotation' issue, and now I can write bad code/mismatched quotes without any warnings, until I run the code and it fails. But as you say, writing sniffs for every botched piece of code a developer can write, is near impossible. What program is actually generating these error messages?

- Mostly, I don't understand why linter-phpcs is not catching these errors and just displaying them with the rest, as per normal operation of the plugin. Is catching syntax errors/illegal code not part of the linter-phpcs functionality?

Bottom line - if I write code with bad syntax, that hasn't yet been 'accounted for' in a given sniff, I'm going to get these errors, so just close them and move on?

Anyway I've spent far too much time on this but I just wanted to say how much I appreciate your time and patience in trying to help me with it. Thank you.

**Comment by gsherwood** [comment]
> Hey Greg, I ran all of the standards over that file and received similar results to yours, so I have no idea what else could be causing that error.

Yeah, that's strange. It's almost like the linter is using a different version of PHPCS, but that's highly unlikely. The linter would be passing content into PHPCS in a different way (via STDIN) but I also tried that with your file during my testing and it worked fine. So I've got no ideas either.

> Bottom line - if I write code with bad syntax, that hasn't yet been 'accounted for' in a given sniff, I'm going to get these errors, so just close them and move on?

I guess so. I have no involvement with the developer working on linter-phpcs, but it might be worth suggesting they do a basic php lint first to ensure there are no syntax errors before continuing. Or, catch the big PHPCS error and just don't display anything. Those big red error boxes certainly do make it seem like something bad has gone wrong, but maybe that is the only feedback available to Atom plugins. I'm not really sure.

> Anyway I've spent far too much time on this but I just wanted to say how much I appreciate your time and patience in trying to help me with it. Thank you.

Not a problem. Hopefully you'll get a setup working eventually, or at least continue to use PHPCS on the command line. Also remember that PHPCS comes with a PHPCBF tool to fix many errors automatically, so you can always run PHPCBF over your files and watch them reload in Atom with a lot of the coding standard errors fixed. I used to use Atom and used to do this myself, but I've since switched away from it because it was crashing too often for me. I used the linter-phpcs plugin for a while there as well, but the constant error checking got a bit annoying for me personally.
